backend/src/main/java/com/google/minicrm/data/Settings.java

```diff
@@ -47,10 +87,27 @@ public Settings(Entity entity) {
         this.phone = (String) entity.getProperty("phone");
         this.phoneNotificationsFrequency =
                 NotificationsFrequency.fromDisplayed((String) entity.getProperty("phoneNf"));
-        this.language = Language.fromIsoCode((String) entity.getProperty("language"));
         this.currency = Currency.fromIsoCode((String) entity.getProperty("currency"));
     }
 
+    public static Settings fromReader(Reader reader, Key advertiserKey) throws NumberParseException, ValidatorException, MalformedJsonException {
+        Gson gson = new Gson();
+        Map<String, String> jsonMap = gson.fromJson(reader, Map.class);
+        if(jsonMap == null || !jsonMap.containsKey("email") || !jsonMap.containsKey("phone") ||
+                !jsonMap.containsKey("emailNotificationsFrequency") ||
+                !jsonMap.containsKey("phoneNotificationsFrequency") ||
+                !jsonMap.containsKey("currency")) {
+            throw new MalformedJsonException("One or more missing keys in JSON.");
+        }
+        String email = jsonMap.get("email");
+        String phone = jsonMap.get("phone");
+        NotificationsFrequency emailNotifFreq = NotificationsFrequency.valueOf(jsonMap.get("emailNotificationsFrequency"));
+        NotificationsFrequency phoneNotifFreq = NotificationsFrequency.valueOf(jsonMap.get("phoneNotificationsFrequency"));
+        Language lang = Language.fromIsoCode(jsonMap.get("language"));
+        Currency curr = Currency.fromIsoCode(jsonMap.get("currency"));
+        return new Settings(advertiserKey, email, phone, emailNotifFreq, phoneNotifFreq, curr);
```

Comment:
it would be nice to refactor this constructor to pass in a value object that will hold all of these fields, as there are probably too many params being passed at this point.

Reply:
I'm a bit confused by what you mean by value object? Are you saying create a builder?

Reply:
just something that can hold all of these values, so instead of passing in a bunch of params, instead you pass in a single object that contains all of these fields. It's not too big of a deal though if it can't be implemented now.

Reply:
Oh I see. But wouldn't that be redundant, since I would have to create a constructor for that value object with all of these fields anyways?

Reply:
oh sorry, yeah you're right about trying to use a Builder, or use @Autovalue for your value class to make it a little easier, example here: https://www.baeldung.com/introduction-to-autovalue. I don't think it's a big deal though.